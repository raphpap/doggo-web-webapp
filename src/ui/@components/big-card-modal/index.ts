export * from './big-card-modal';
export {default} from './big-card-modal';
