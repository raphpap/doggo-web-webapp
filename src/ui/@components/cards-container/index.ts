export * from './card-container';
export {default} from './card-container';
