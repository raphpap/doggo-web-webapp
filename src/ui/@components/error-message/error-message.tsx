// Vendor
import React from 'react';
import styled from 'react-emotion';

// Theme
import theme from 'doggo-web-webapp/theme';

interface Props {
  message: string;
}

const Container = styled.div`
  display: flex;
  opacity: 0.8;
  align-items: center;
  margin: ${theme.margin.unit * 1.5}px 0 ${theme.margin.unit}px;
  font-size: 13px;
  line-height: 15px;
  color: ${theme.colors.white.plain};
`;

export const ErrorMessage: React.SFC<Props> = ({message}) => (
  <Container>
    <span>{message}</span>
  </Container>
);

export default ErrorMessage;
