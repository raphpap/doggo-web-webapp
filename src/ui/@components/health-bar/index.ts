export * from './health-bar';
export {default} from './health-bar';
