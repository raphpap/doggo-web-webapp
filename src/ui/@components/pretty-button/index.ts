export * from './pretty-button';
export {default} from './pretty-button';
