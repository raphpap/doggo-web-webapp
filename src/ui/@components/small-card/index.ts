export * from './small-card';
export {default} from './small-card';
