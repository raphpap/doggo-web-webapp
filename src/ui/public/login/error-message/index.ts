export * from './error-message';
export {default} from './error-message';
