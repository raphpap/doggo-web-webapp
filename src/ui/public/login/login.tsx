// Vendor
import React, {ChangeEventHandler, FormEventHandler} from 'react';
import {compose} from 'recompose';

// Vendor components
import {Redirect, RouteComponentProps, withRouter} from 'react-router-dom';

// Assets
import logo from 'doggo/assets/images/doggo-logo.png';

// Shared components
import ErrorMessage from 'doggo/ui/@components/error-message';
import Logo from 'doggo/ui/@components/logo';

// Components
import Button from './button';
import Form from './form';
import TextField from './text-field';

// Context
import {
  withApplicationContext,
  WithApplicationContextProps
} from 'doggo/context';

// Types
interface State {
  username: string;
  password: string;
  isDirty: boolean;
}

interface Props extends RouteComponentProps<never> {}
type EnhancedProps = Props & WithApplicationContextProps;

const enhance = compose<EnhancedProps, Props>(
  withApplicationContext,
  withRouter
);

export class Login extends React.Component<EnhancedProps, State> {
  public readonly state: Readonly<State> = {
    isDirty: false,
    password: '',
    username: ''
  };

  public render() {
    const {cards, error} = this.props.context.state;
    const isAuthenticated = !!cards;
    const {from} = this.props.location.state || {from: {pathname: '/capture'}};

    if (isAuthenticated) return <Redirect to={from} />;

    const hasError = Boolean(error);
    const {isDirty, username, password} = this.state;

    return (
      <>
        <Logo logo={logo} marginUnits={5} height={'180px'} />

        <Form onSubmit={this.handleSubmit}>
          <TextField
            id="username"
            error={!isDirty && hasError}
            placeholder="Username"
            type="text"
            value={username}
            onChange={this.handleUsername}
          />

          <TextField
            id="password"
            error={!isDirty && hasError}
            placeholder={'Password'}
            type="password"
            value={password}
            onChange={this.handlePassword}
          />

          {!isDirty && hasError && (
            <ErrorMessage message="Wrong username and/or password" />
          )}

          <Button type="submit">Login</Button>
        </Form>
      </>
    );
  }

  private handleUsername: ChangeEventHandler<HTMLInputElement> = event => {
    this.setState({
      isDirty: true,
      username: event.target.value
    });
  };

  private handlePassword: ChangeEventHandler<HTMLInputElement> = event => {
    this.setState({
      isDirty: true,
      password: event.target.value
    });
  };

  private handleSubmit: FormEventHandler<HTMLFormElement> = async event => {
    event.preventDefault();

    const {login} = this.props.context.actions;
    const {username, password} = this.state;

    login(username, password);
    this.setState({isDirty: false});
  };
}

export default enhance(Login);
