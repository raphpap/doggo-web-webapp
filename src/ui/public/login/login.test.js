// Vendor
import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

// Vendor components
import {Redirect} from 'react-router-dom';

// Shared Components
import {ErrorMessage} from 'doggo/ui/@components/error-message';

// Components
import {Login} from './login';
import TextField from './text-field';

// Mocks
import {mockContextWithoutData, mockLoggedIn} from 'doggo/test-mocks';

describe('<Login />', () => {
  let wrapper;
  let context;
  let location = {state: null};

  beforeEach(() => {
    context = mockContextWithoutData();
    wrapper = shallow(<Login context={context} location={location} />);
  });

  it('renders correctly when not logged in', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('redirects to /capture after logged in when no previous path', () => {
    context = mockLoggedIn();
    wrapper.setProps({context});
    const redirect = wrapper.find(Redirect);
    expect(redirect.props().to.pathname).toBe('/capture');
  });

  it('redirects to the right path after logged in with previous path', () => {
    context = mockLoggedIn();
    location = {state: {from: {pathname: '/previous-path'}}};
    wrapper.setProps({context, location});
    const redirect = wrapper.find(Redirect);
    expect(redirect.props().to.pathname).toBe('/previous-path');
  });

  it('shows the errors when login fail', () => {
    context.state.error = {status: 401};
    wrapper.setProps({context: context});

    const errorMessage = wrapper.find(ErrorMessage);
    const textFields = wrapper.find(TextField);

    expect(textFields.at(0).props().error).toBe(true);
    expect(textFields.at(1).props().error).toBe(true);
    expect(errorMessage).toHaveLength(1);
  });
});
