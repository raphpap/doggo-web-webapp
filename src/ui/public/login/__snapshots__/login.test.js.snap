// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> renders correctly when not logged in 1`] = `
<Fragment>
  <Logo
    height="180px"
    logo="doggo-logo.png"
    marginUnits={5}
  />
  <Form
    onSubmit={[Function]}
  >
    <TextField
      error={false}
      id="username"
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value=""
    />
    <TextField
      error={false}
      id="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <Button
      type="submit"
    >
      Login
    </Button>
  </Form>
</Fragment>
`;
